### PR TITLE
Implement IndexOrDocValuesQuery for geo_shape field

### PR DESCRIPTION
**Pull request by iverase**
IndexOrDocValuesQuery is an optimisation for queries when there are in a filter with another query that is more restrictive. In that case, it might be faster to just check the matches for that query directly with the doc values instead of traversing the BKD index.

With the introduction of doc values for geo_shape, it is possible now to implement  such query for this type of data.

### CODE REVIEW

File: x-pack/plugin/spatial/src/test/java/org/elasticsearch/xpack/spatial/index/query/LatLonShapeDocValuesQueryTests.java [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+            Query indexQuery = LatLonShape.newGeometryQuery(FIELD_NAME, ShapeField.QueryRelation.INTERSECTS, q);
+            Query docValQuery = new LatLonShapeDocValuesQuery(FIELD_NAME,ShapeField.QueryRelation.INTERSECTS, q);
+            assertEquals(s.count(indexQuery), s.count(docValQuery));
+            indexQuery = LatLonShape.newGeometryQuery(FIELD_NAME, ShapeField.QueryRelation.WITHIN, q);
+            docValQuery = new LatLonShapeDocValuesQuery(FIELD_NAME, ShapeField.QueryRelation.WITHIN, q);
+            assertEquals(s.count(indexQuery), s.count(docValQuery));
+            indexQuery = LatLonShape.newGeometryQuery(FIELD_NAME, ShapeField.QueryRelation.DISJOINT, q);
+            docValQuery = new LatLonShapeDocValuesQuery(FIELD_NAME, ShapeField.QueryRelation.DISJOINT, q);
+            assertEquals(s.count(indexQuery), s.count(docValQuery));
+            // CONTAINS and multi-shapes fails due to LUCENE-9595
```
Comment (jpountz) [inline x-pack/plugin/spatial/src/test/java/org/elasticsearch/xpack/spatial/index/query/LatLonShapeDocValuesQueryTests.java:162]: do you think it should block this PR?

File: x-pack/plugin/spatial/src/test/java/org/elasticsearch/xpack/spatial/index/query/LatLonShapeDocValuesQueryTests.java [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+                List<IndexableField> fields = indexer.indexShape(null, geometry);
+                for (IndexableField field : fields) {
+                    doc.add(field);
+                }
+                BinaryGeoShapeDocValuesField docVal = new BinaryGeoShapeDocValuesField(FIELD_NAME);
+                docVal.add(fields, geometry);
+                doc.add(docVal);
+                w.addDocument(doc);
+            } catch (Exception e) {
+                // ignore
```
Comment (jpountz) [inline x-pack/plugin/spatial/src/test/java/org/elasticsearch/xpack/spatial/index/query/LatLonShapeDocValuesQueryTests.java:140]: is it safe? Can we narrow down the set of exceptions types we ignore?

File: x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/index/query/LatLonShapeDocValuesQuery.java [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+
+    private final String field;
+    private final LatLonGeometry[] geometry;
+    private final ShapeField.QueryRelation relation;
+
+    LatLonShapeDocValuesQuery(String field, ShapeField.QueryRelation relation, LatLonGeometry... geometry) {
+        if (field == null) {
+            throw new IllegalArgumentException("field must not be null");
+        }
+        this.field = field;
```
Comment (jpountz) [inline x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/index/query/LatLonShapeDocValuesQuery.java:42]: should you also test that `geometry` has a length > 0?

Comment (jpountz) [reply x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/index/query/LatLonShapeDocValuesQuery.java:42]: or alternatively rewrite the query when length == 0?

**Review by jpountz — COMMENTED** [review]
I didn't review the Component2DVisitor logic, but the way the query gets built makes sense to me.

**Review by talevy — COMMENTED** [review]
Do you think it would be helpful to have both index and doc-value queries execute against the same tests instead of just  having doc-value queries be considered correct if they behave like index-queries? Since theoretically the counts may be the same, but maybe they are including different documents? highly unlikely, but just a thought.

**Comment by iverase** [comment]
Thanks @Tal. I updated the test so we actually check the hits are the same. I improved the test as well which surface and issue with Contains queries and geometry collections.

I added more documentation to Component2DVisitor, let me know if that's ok.

File: x-pack/plugin/spatial/src/test/java/org/elasticsearch/xpack/spatial/index/query/LatLonShapeDocValuesQueryTests.java (hunk 2 above)
Comment (iverase) [reply x-pack/plugin/spatial/src/test/java/org/elasticsearch/xpack/spatial/index/query/LatLonShapeDocValuesQueryTests.java]: I remove the try / catch

File: x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/index/query/LatLonShapeDocValuesQuery.java (hunk 3 above)
Comment (iverase) [reply x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/index/query/LatLonShapeDocValuesQuery.java:45]: This is done in `VectorGeoShapeWithDocValuesQueryProcessor`. It gets rewritten to a `MatchNoDocsQuery`, is that enough?

File: x-pack/plugin/spatial/src/test/java/org/elasticsearch/xpack/spatial/index/query/LatLonShapeDocValuesQueryTests.java (hunk 1 above)
Comment (iverase) [reply x-pack/plugin/spatial/src/test/java/org/elasticsearch/xpack/spatial/index/query/LatLonShapeDocValuesQueryTests.java]: I think we can wait until we depend on a Lucene version that contains the fix.

**Comment by iverase** [comment]
We upgraded Elasticsearch to a Lucene 8.8 snapshot so this PR is not blocked any more. Could you have another look @talevy ?

File: x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/index/query/LatLonShapeDocValuesQuery.java [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+                    public boolean matches() throws IOException {
+                        reader.reset(values.binaryValue());
+                        visitor.reset();
+                        reader.visit(visitor);
+                        return visitor.matches();
+                    }
+
+                    @Override
+                    public float matchCost() {
+                        return 1000f; // TODO: what should it be?
```
Comment (talevy) [inline x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/index/query/LatLonShapeDocValuesQuery.java:122]: I am not too familiar with this. So long as `matchCost` is constant and the ConstantScoreScorer is used, this number is arbitrary and has no effect, right?

Comment (iverase) [reply x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/index/query/LatLonShapeDocValuesQuery.java:122]: My understanding is that this value is used when you have more that one `TwoPhaseIterator` in a conjunction in order to sort them. You want to execute the cheaper iterators first.

In put case we do not have a good estimation of cost so it is good enough to return a big number. Note that this is done the same for LatLonPointDocValues.

Comment (talevy) [reply x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/index/query/LatLonShapeDocValuesQuery.java:122]: thanks!

**Review by talevy — APPROVED** [review]
LGTM pending resolution

**Pull request merged**